### ISSUE TITLE
HTML frame content size is determined by Layout Export Resolution

### ISSUE THREAD

**Issue by baswein** [qgis/QGIS#37864]
**Describe the bug**

This was a  fun one to get to the bottom of :) The size of the font when linking an html file using an HTML Frame Item is determined by the export resolution of the Layout. The default resolution of 300 dpi is close (but not perfect) to the expected size.    
**How to Reproduce**
Open up a new layout with the export resolution set to 300 dpi and insert a HTML Frame Item and link to an HTML file. Now change the export resolution of the layout to 100 dpi and insert the same HTML file. This one will be 3 times as large as the last one. 
![HTML_Insert_size](https://user-images.githubusercontent.com/15817745/87823936-59e81480-c842-11ea-874d-8374141823a3.gif)
[test_arial_code.zip](https://github.com/qgis/QGIS/files/4939822/test_arial_code.zip)

**QGIS and OS versions**
Also tried in 3.10.7
QGIS version
3.14.0-Pi
QGIS code revision
9f7028fd23

Compiled against Qt
5.11.2
Running against Qt
5.11.2
Compiled against GDAL/OGR
3.0.4
Running against GDAL/OGR
3.0.4
Compiled against GEOS
3.8.1-CAPI-1.13.3
Running against GEOS
3.8.1-CAPI-1.13.3
Compiled against SQLite
3.29.0
Running against SQLite
3.29.0
PostgreSQL Client Version
11.5
SpatiaLite Version
4.3.0
QWT Version
6.1.3
QScintilla2 Version
2.10.8
Compiled against PROJ
6.3.2
Running against PROJ
Rel. 6.3.2, May 1st, 2020
OS Version
Windows 10 (10.0)
Active python plugins
db_manager; 
MetaSearch; 
processing

**Additional context**
I suppose this could be considered a feature of it was well documented.

**Comment by nyalldawson** [comment]
I don't see this as an issue -- should we move it to the documentation issue tracker?

**Comment by baswein** [comment]
As a short term fix that could work. 
It's not a terrible issue if you know what is happening. Other wise it is a frustrating and perplexing phenomenon. So I would think a note would need to be included in the HTML source properties. Maybe something like this. 
![image](https://user-images.githubusercontent.com/15817745/88085187-8312e880-cb53-11ea-89da-9c5b3e9e9b3f.png)
Long term don't you think it would be better to have a scale factor in the HTML Source properties?  
I did some experiments and the scale is set by the output dpi when the html frame is created not when the html URL is linked.  
Also as far as I can tell to match text sizes between a QGIS label and a HTML frame the export resolution needs to be somewhere between 271 and 272 dpi. 
I am excited about the HTML frames because I have found that using ms word (sorry libre office messes up the formatting as far as I can tell) to create and edit HTML files we can have WYSIWYG complex formatting in layouts that anyone can edit.   
Of course a WYSIWYG editor in QGIS is still a dream.

**Comment by christophfink** [comment]
I am facing this issue when trying to scale text using CSS, for print. In this use case it is a serious issue to not be able to provide precise font sizes and still export in reasonable resolution to PDF. Note that CSS allows to specify dimensions in units that are (ought to be) independent from screen/print resolution, e.g. pt and mm. 
The workaround of using “Label” text boxes only does not work in my case, because they lack other important layout tools, such as e.g. specifying line spacing. 

I’m trying to argue against this being a mere issue of proper documentation. The HTML text boxes, as of now, are basically the only way of properly formatting text - measurements within them should not change arbitrarily.

**Comment by baswein** [comment]
Hello friendly closing robot I would like to plead my case to your tinny ears that this fine issue deserves to live to see satisfaction as it is a valid issue with a very hacky work around.

**Comment by gioman** [comment]
> I don't see this as an issue

@baswein I'll leave to a developer (@nyalldawson)  to choose if this must be closed (as issue) and instead be documented.

**Comment by baswein** [comment]
Would an appropriate resolution be to document the current behavior and make this a feature request for a scale factor in the HTML Source properties?

**Comment by gioman** [comment]
> Would an appropriate resolution be to document the current behavior and make this a feature request for a scale factor in the HTML Source properties?

@baswein document it https://github.com/qgis/QGIS/issues/37864#issuecomment-661546361

**Comment by DelazJ** [comment]
Is this really a documentation(-only?) issue? According to the description, it looks hard to get the imports scale at 100% without doing some math (and nothing obviously informs/helps in the GUI).

**Comment by baswein** [comment]
@DelazJ I agree with you that it is not a documentation only issue but it really doesn't require math for most purposes. If you want the want the font size to be imperceptibly close (.05mm over 100mm) to what is says in the HTML code all you need to do is set the layout export resolution to 272dpi when creating the HTML frame and then switch it back to your desired export resolution when finished. 

If you want it to be exactly the same size this is not possible because it shifts when changing font sizes, but not that much, and the exact value can be between dpi integers.

**Comment by nyalldawson** [comment]
What about if we add a dedicated property for each html item to allow the dpi to be set for the item?

**Comment by baswein** [comment]
That sounds like a fine solution to me. 
However I think a Scale property would be clearer than a dpi property. 
The dpi relationship is still confusing to me and I assume would be confusing to other users. 
I would rather see a dedicated property that was the scale of the html item so that 100% would make a font that is set at 11pt by the html code be the same size as a font set to 11pt in a label regardless of the output resolution. Unless I am missing something this is how it works in other aspects of QGIS except for vector and raster tiles. 

Is dpi important for html items that are not text? 
Do things like css absolute lengths, em, ex, etc. mess with that?   

Thanks again for looking into this and everything you do! I can't even find where this is set in the .qgs file.

**Comment by Pedro-Murteira** [comment]
I can still reproduce this issue on QGIS 3.22.4 and 3.24.0.

**Comment by baswein** [comment]
I didn't realize that the HTML changes on restart of the project. That makes things really annoying!
